Add Reader Featured image alt text

ivan-ottinger opened the pull request:
#### Changes proposed in this Pull Request

The proposed code adds alt text to post Featured image in Reader at `read/feeds/[xxxxxxxx]/posts/[xxxxxxxxx]`:

![Markup on 2021-08-11 at 14:30:23](https://user-images.githubusercontent.com/25105483/129028898-b16065bc-5bf8-494b-b20f-47e948e70ff5.png)

The code tries to extract the alt text from `post.attachments` object first. If there is no post attachment available (e.g. when the Featured image was uploaded _before_ the time when the post was created) - OR - if the existing post attachment doesn't have any alt text, the post title will be used as the alt text instead.

#### Testing instructions

1. Create a new post with Featured image.
2. Add alt text to the Featured image through Media section in Calypso.
3. Navigate to Reader at https://wordpress.com/read.
4. Open the new post in Reader. You should be at `read/feeds/[xxxxxxxx]/posts/[xxxxxxxxx]` (not on the post page itself).
5. Open the Console and take a look if the Featured image has its alt text.
6. Try to remove alt text from the image through the Media section and go back to `read/feeds/[xxxxxxxx]/posts/[xxxxxxxxx]`.
7. When you open the Console this time, the alt text of the Featured image should be the same as the site title.

Related to https://github.com/Automattic/wp-calypso/issues/49226

ivan-ottinger committed:
Add Featured image alt text in Reader

This commit adds alt text to Featured images displayed in Reader.

The code tries to extract the alt text from `post.attachments`  first.

If there are no post attachments available, the post title will be used as the alt text instead.

ivan-ottinger committed:
Fix `getFeaturedImageAlt()` parameter description

alshakero commented on client/blocks/reader-full-post/featured-image.jsx:
```diff
@@ -23,12 +29,13 @@ export default class FeaturedImage extends React.Component {
 		return (
 			<div className="reader-full-post__featured-image">
-				<img src={ this.state.src } onError={ this.handleImageError } alt="" />
+				<img src={ this.state.src } onError={ this.handleImageError } alt={ this.state.alt } />
```
Why not use `this.props.alt` without storing it? This is a classic React anti-pattern.

https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#anti-pattern-unconditionally-copying-props-to-state

ivan-ottinger replied:
Thank you very much! This is now fixed in https://github.com/Automattic/wp-calypso/pull/55378/commits/f8c9e46a0266859e584255db16a79e711def9cff.

alshakero commented on client/reader/get-helpers.js:
```diff
@@ -160,3 +160,27 @@ export const canBeMarkedAsSeen = ( { post = null, posts = [], currentRoute = ''
+export const getFeaturedImageAlt = ( post ) => {
+	// Each post can have multiple images attached. To make sure we are selecting
+	// the alt text of the correct image attachment, we get the ID of the post thumbnail first
+	// and then use it to get the alt text of the Featured image.
+	const postThumbnailId = get( post, 'post_thumbnail.ID' );
+	const featuredImageAlt = get( post, 'attachments.' + postThumbnailId + '.alt' );
+	const postTitle = post.title;
```
In an attempt to phase out lodash whenever we can, we can do

```suggestion
	// and then use it to get the alt text of the Featured image.
	const postTitle = post.title;
	const postThumbnailId = post?.post_thumbnail?.ID;
	const featuredImageAlt = post?.attachments[postThumbnailId]?.alt;	
```

ivan-ottinger replied:
> In an attempt to phase out lodash whenever we can, we can do

Thanks Omar! That's a great TIL. I felt that using `get()` was a bit clumsy, but didn't know about `?.`.

I have fixed that now in the following commit: https://github.com/Automattic/wp-calypso/pull/55378/commits/fb317dced040699bb86ea2412e536dd001751adc

alshakero reviewed (commented):
I left two minor comments. But honestly, really beautiful and clean work.

ivan-ottinger committed:
Replace usage of `get()` (lodash) with `?.`

ivan-ottinger committed:
Replace `this.state.alt` with `this.props.alt`

The proposed change makes sure that the alt text is passed directly via component prop (instead of its state).

arcangelini commented on client/reader/get-helpers.js:
```diff
@@ -160,3 +160,27 @@ export const canBeMarkedAsSeen = ( { post = null, posts = [], currentRoute = ''
+	// If there is no Featured image alt text available, return post title instead.
+	// This will make sure that the featured image has at least some relevant alt text.
+	if ( ! featuredImageAlt ) {
+		return postTitle;
```
This is just a personal opinion, but as I see it since the feed is not a public-facing thing the ALT is less about SEO and more about accessibility. With that in mind, it could be helpful to set it as `postTitle + ' featured image'` to be a bit more descriptive?

What you have is great as well. Just wanted to throw my opinion out there.

yansern replied:
Feel free to pick and mix your favorite ice cream flavours.
I'm not sure which package is preferable.

```js
// class components using @wordpress package
import { __ } from '@wordpress/i18n';
import { sprintf } from '@wordpress/i18n';

// class components using calypso package
import { translate } from 'i18n-calypso';

// function components using @wordpress package
import { useI18n } from '@wordpress/react-i18n';
const { __ } = useI18n();

// function components using calypso package
import { useTranslate } from 'i18n-calypso';
const translate = useTranslate();

// translators: explain what this string is for
sprintf( __( 'featured image of the post %s' ), 'post title here'} );

// translators: explain what this string is for
sprintf( __( 'featured image of the post %(postTitle)s and here is some numbers %(age)d' ), {
    postTitle: 'post title here',
    age: 32
 } );
 
 // translators: explain what this string is for
translate( 'featured image of the post %(postTitle)s and here is some numbers %(age)d', {
    postTitle: 'post title here',
    age: 32
 } );
```

ivan-ottinger replied:
Thank you for your suggestions, folks!

I have tested the experience using screen reader and it seems to be better with adding the "featured image" context to the alt attribute.

New commit: https://github.com/Automattic/wp-calypso/pull/55378/commits/1541ec6b591627d0f89576cb3558b3a0cc357c4e

> Feel free to pick and mix your favorite ice cream flavours.
I'm not sure which package is preferable.

I have relied on 

```javascript
// class components using calypso package
import { translate } from 'i18n-calypso'; 
```

since it was already used in the same file.

arcangelini reviewed (approved):
LGTM, just had that one thought on the ALT text for posts without an attachment.

ivan-ottinger committed:
Add `featured image` description to the alt text

ivan-ottinger commented:
PR deployed to production.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6409689

Thank you @ivan-ottinger for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6409689).
